lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3875,25 +3904,36 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   /// The parent VJP emitter.
   VJPEmitter &vjpEmitter;
 
-  /// Dominance info for the original function.
-  DominanceInfo *domInfo;
-
-  /// Post-dominance info for the original function.
-  PostDominanceInfo *postDomInfo;
+  /// Post-order info for the original function.
+  PostOrderFunctionInfo *postOrderInfo = nullptr;
 
   /// Mapping from original values to their corresponding adjoint values.
-  DenseMap<SILValue, AdjointValue> valueMap;
+  // DenseMap<SILValue, AdjointValue> valueMap;
+  DenseMap<SILBasicBlock *, DenseMap<SILValue, AdjointValue>> valueMap;
 
   /// Mapping from original buffers to their corresponding adjoint buffers.
   DenseMap<SILValue, ValueWithCleanup> bufferMap;
 
-  /// Mapping from original basic blocks to their corresponding adjoint basic
-  /// blocks.
+  /// Mapping from original basic blocks to corresponding adjoint basic blocks.
+  /// Adjoint basic blocks always have the predecessor as the single argument.
   DenseMap<SILBasicBlock *, SILBasicBlock *> adjointBBMap;
 
-  /// Local stack allocations.
+  /// Mapping from adjoint basic blocks to pullback struct arguments.
+  DenseMap<SILBasicBlock *, SILArgument *> adjointPullbackStructArguments;
+
+  /// Mapping from original basic blocks and successor basic blocks to
+  /// corresponding adjoint forwarding basic blocks. Forwarding basic blocks
+  /// take additional arguments in addition to the predecessor enum argument.
+  DenseMap<std::pair<SILBasicBlock *, SILBasicBlock *>, SILBasicBlock *>
+      adjointForwardingBBMap;
```

Comment:
Rename to `adjointTrampolineBBMap`?
"Trampoline" is standard terminology for "basic blocks that just branch" - see `SILBasicBlock::isTrampoline()`.
Though adjoint trampoline BBs may also materialize `AdjointValue`s to `SILValue`s before branching.

Reply:
Yes

Reply:
Done.